File: src/Cpp/17_Slowniki/Zad11.cpp
#include <algorithm>
#include <cassert>
#include <unordered_map>
#include <utility>
#include <vector>

//Otrzymujesz dwie listy. Zbuduj slownik skladajacy się z kluczy będacych
//elementami pierwszej listy i wartosciami będacymi elementami drugiej listy.
//Jesli listy sa roznej dlugosci zwroc pusty slownik.

bool porownujKlucze(const std::pair<std::string, int>& a, const std::pair<std::string, int>& b)
{
    return a.first < b.first;
}

std::vector<std::pair<std::string, int>> sortujWzgledemKluczy(const std::unordered_map<std::string, int>& slownik)
{

    std::vector<std::pair<std::string, int>> wynik;

    for (const auto& it : slownik)
        wynik.push_back(std::pair<std::string, int>(it.first, it.second));

    sort(wynik.begin(), wynik.end(), porownujKlucze);

    return wynik;
}

bool porownujWartosci(const std::pair<std::string, int>& a, const std::pair<std::string, int>& b)
{
    return a.second < b.second;
}

std::vector<std::pair<std::string, int>> sortujWzgledemWartosci(const std::unordered_map<std::string, int>& slownik)
{

    std::vector<std::pair<std::string, int>> wynik;

    for (const auto& it : slownik)
        wynik.push_back(std::pair<std::string, int>(it.first, it.second));

    sort(wynik.begin(), wynik.end(), porownujWartosci);

    return wynik;
}

void test1()
{
    std::unordered_map<std::string, int> slownik{ { "c", 3 }, { "x", 5 }, { "a", -2 }, { "b", 4 } };
    std::vector<std::pair<std::string, int>> wynik{ { "a", -2 }, { "c", 3 }, { "b", 4 }, { "x", 5 } };

    assert(sortujWzgledemWartosci(slownik) == wynik);
}

void test2()
{
    std::unordered_map<std::string, int> slownik{ { "c", 3 }, { "x", 5 }, { "a", -2 }, { "b", 4 } };
    std::vector<std::pair<std::string, int>> wynik{ { "a", -2 }, { "b", 4 }, { "c", 3 }, { "x", 5 } };

    assert(sortujWzgledemKluczy(slownik) == wynik);
}

int main()
{

    test1();
    test2();

    return 0;
}

File: src/JavaScript/17_Slowniki/Zad11.js
/*
Otrzymujesz slownik skladajacy sie z par napisow i liczb calkowitych.
a) Wypisz pary posortowane wzgledem napisow.
b) Wypisz pary posortowane wzgledem liczb.
*/
var Pair = /** @class */ (function() {
    function Pair(a, b) {
        this.first = a;
        this.second = b;
    }
    return Pair;
}());

sortujWzgledemNapisow = function(slownik) {
    var wynik = [];
    for (const key in slownik) {
        wynik.push(new Pair(key, slownik[key]));
    }

    wynik.sort(function(a, b) {
            if (a.first < b.first) {
                return -1;
            }
            if (a.first > b.first) {
                return 1;
            }
            return 0;
        }

    );
    return wynik;
};


sortujWzgledemWartosci = function(slownik) {
    var wynik = [];
    for (const key in slownik) {
        wynik.push(new Pair(key, slownik[key]));
    }

    wynik.sort(function(a, b) {
            if (a.second < b.second) {
                return -1;
            }
            if (a.second > b.second) {
                return 1;
            }
            return 0;
        }

    );
    return wynik;
}

test1 = function() {
    var slownik = {
        "c": 3,
        "x": 5,
        "a": -2,
        "b": 4
    };
    var oczekiwane = [new Pair("a", -2), new Pair("c", 3), new Pair("b", 4), new Pair("x", 5)];
    var wynik = sortujWzgledemWartosci(slownik);

    if (!(wynik.length == oczekiwane.length)) {
        throw new Error(`Assertion error line 34: ${wynik.length} != ${oczekiwane.length}`);
    }

    for (var i = 0; i < wynik.length; i++) {
        if (!(wynik[i].first == oczekiwane[i].first && wynik[i].second == oczekiwane[i].second)) {
            throw new Error(`Assertion error line 38: ${wynik[i].first} != ${oczekiwane[i].first} || ${wynik[i].second} != ${oczekiwane[i].second}`);
        }
    }
};

test2 = function() {
    var slownik = {
        "c": 3,
        "x": 5,
        "a": -2,
        "b": 4
    };
    var oczekiwane = [new Pair("a", -2), new Pair("b", 4), new Pair("c", 3), new Pair("x", 5)];
    var wynik = sortujWzgledemNapisow(slownik);

    if (!(wynik.length == oczekiwane.length)) {
        throw new Error(`Assertion error line 34: ${wynik.length} != ${oczekiwane.length}`);
    }

    for (var i = 0; i < wynik.length; i++) {
        if (!(wynik[i].first == oczekiwane[i].first && wynik[i].second == oczekiwane[i].second)) {
            throw new Error(`Assertion error line 38: ${wynik[i].first} != ${oczekiwane[i].first} || ${wynik[i].second} != ${oczekiwane[i].second}`);
        }
    }
}

main = function() {
    test1();
    test2();
}

main();